src/main/java/com/google/sps/DataServlet.java

```diff
@@ -164,37 +201,59 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-    if (nodeNameParam.length() > 0) {
-      queried.add(nodeNameParam);
-    }
-    // If the token is contained, then get the nodes associated with the token and
-    // add them to the queried nodes
+    // Show mutations relevant to nodes that contain the token in the new graph
     if (currDataGraph.tokenMap().containsKey(tokenParam)) {
       queried.addAll(currDataGraph.tokenMap().get(tokenParam));
+      queriedNext.addAll(currDataGraph.tokenMap().get(tokenParam));
     }
+
+    // Truncate the graph from the nodes that the client had searched for
     truncatedGraph = currDataGraph.getReachableNodes(queried, depthNumber);
 
-    if (nodeNameParam.length() == 0 && truncatedGraph.equals(currDataGraph.graph())) {
-      // If we are not filtering the graph or limiting its depth, show all mutations of all nodes
+    // To get the nodes to calculate relevant mutations from. If queried and queried next contain
+    // the same
+    // nodes, then no reason to regenerate the graph
+    MutableGraph<GraphNode> truncatedGraphNext =
+        queried.equals(queriedNext)
+            ? Graphs.copyOf(truncatedGraph)
+            : currDataGraph.getReachableNodes(queriedNext, depthNumber);
+
+    // If we are not filtering the graph or limiting its depth, show all mutations of all nodes
+    if (nodeNames.size() == 0
+        && tokenParam.length() == 0
+        && truncatedGraph.equals(currDataGraph.graph())) {
       filteredMutationIndices = defaultIndices;
+      filteredDiff = diff;
     } else {
       // Get the names of all the displayed nodes and find all indices of mutations
       // that mutate any of them
       Set<String> truncatedGraphNodeNames = Utility.getNodeNamesInGraph(truncatedGraph);
+      Set<String> truncatedGraphNodeNamesNext = Utility.getNodeNamesInGraph(truncatedGraphNext);
 
       // Also get mutations relevant to the searched node if it is not an empty string
-      if (nodeNameParam.length() != 0) {
-        truncatedGraphNodeNames.add(nodeNameParam);
+      if (nodeNames.size() != 0) {
+        truncatedGraphNodeNames.addAll(nodeNames);
       }
 
-      filteredMutationIndices =
-          Utility.findRelevantMutations(truncatedGraphNodeNames, mutationIndicesMap, mutList);
-
-      // Filter the diff to only show mutations relevant to the above nodes
-      diff = Utility.filterMultiMutationByNodes(diff, truncatedGraphNodeNames);
+      // A set containing a indices where nodes currently displayed on the graph
+      // or queried are mutated
+      Set<Integer> mutationIndicesSet = new HashSet<>();
+
+      mutationIndicesSet.addAll(
+          Utility.findRelevantMutations(truncatedGraphNodeNamesNext, mutationIndicesMap, mutList));
+      mutationIndicesSet.addAll(Utility.getMutationIndicesOfToken(tokenParam, mutList));
+      filteredMutationIndices = new ArrayList<>(mutationIndicesSet);
+      Collections.sort(filteredMutationIndices);
+
+      if (tokenParam.length() == 0) {
```

Comment:
Can we eliminate the check on line 234 by just filtering by the union of the original `truncatedGraphNodeNames` and `queried`? In the case that no token is searched, queried will just contain the node names if that list is non-empty?